gobblin-yarn/src/main/java/gobblin/yarn/YarnService.java

```diff
@@ -0,0 +1,457 @@
+import org.apache.hadoop.yarn.api.records.ContainerStatus;
+import org.apache.hadoop.yarn.api.records.FinalApplicationStatus;
+import org.apache.hadoop.yarn.api.records.LocalResource;
+import org.apache.hadoop.yarn.api.records.LocalResourceType;
+import org.apache.hadoop.yarn.api.records.NodeReport;
+import org.apache.hadoop.yarn.api.records.Priority;
+import org.apache.hadoop.yarn.api.records.Resource;
+import org.apache.hadoop.yarn.client.api.AMRMClient;
+import org.apache.hadoop.yarn.client.api.async.AMRMClientAsync;
+import org.apache.hadoop.yarn.client.api.async.NMClientAsync;
+import org.apache.hadoop.yarn.conf.YarnConfiguration;
+import org.apache.hadoop.yarn.exceptions.YarnException;
+import org.apache.hadoop.yarn.security.AMRMTokenIdentifier;
+import org.apache.hadoop.yarn.util.Records;
+import org.apache.helix.api.id.ParticipantId;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Splitter;
+import com.google.common.base.Optional;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.eventbus.EventBus;
+import com.google.common.eventbus.Subscribe;
+import com.google.common.io.Closer;
+import com.google.common.util.concurrent.AbstractIdleService;
+
+import com.typesafe.config.Config;
+
+import gobblin.util.ExecutorsUtils;
+import gobblin.yarn.event.ApplicationMasterShutdownRequest;
+import gobblin.yarn.event.ContainerShutdownRequest;
+import gobblin.yarn.event.NewContainerRequest;
+
+
+/**
+ * This class is responsible for all Yarn-related stuffs including ApplicationMaster registration,
+ * ApplicationMaster un-registration, Yarn container management, etc.
+ *
+ * @author ynli
+ */
+public class YarnService extends AbstractIdleService {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(YarnService.class);
+
+  private static final Splitter SPLITTER = Splitter.on(',').omitEmptyStrings().trimResults();
+
+  private final ConcurrentMap<ContainerId, Container> containerMap = Maps.newConcurrentMap();
+  private final ConcurrentMap<ContainerId, ParticipantId> containerToParticipantMap = Maps.newConcurrentMap();
+
+  private final ApplicationId applicationId;
+  private final String applicationName;
+
+  private final Config config;
+
+  private final Configuration yarnConfiguration;
+  private final AMRMClientAsync<AMRMClient.ContainerRequest> amrmClientAsync;
+  private final NMClientAsync nmClientAsync;
+  private final ExecutorService containerLaunchExecutor;
+  private final FileSystem fs;
+
+  private final EventBus eventBus;
+
+  private final int initialContainers;
+  private final int requestedContainerMemoryMbs;
+  private final int requestedContainerCores;
+
+  private final String containerJvmArgs;
+
+  // Security tokens for accessing HDFS
+  private final ByteBuffer tokens;
+
+  private final Closer closer = Closer.create();
+
+  private final Object allContainersStopped = new Object();
```

Comment:
We probably don't want to do that because in the future the number of contains will change at runtime and `containerMap.size()` will no longer be a good indication.